### PR TITLE
Use Unsafe.compareAndSwap for writing to Stable field and array

### PR DESCRIPTION
**Pull request by nbhuiyan**
Using Unsafe.compareAndSwap prevents issues resulting from multiple threads coming across uninitialized VarHandle.typesAndInvokers field and competing to populate the field with their own versions. Using compareAndSwap to populate the typesAndInvokers' MethodHandle table also prevents similar issues when multiple threads come across the same unpopulated slot in the table and compete to populate it.

Issue: eclipse-openj9/openj9#18882

### CODE REVIEW

File: src/java.base/share/classes/java/lang/invoke/VarHandle.java [hunk 1]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2014, 2020, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2014, 2024, Oracle and/or its affiliates. All rights reserved.
```
Comment (pshipton) [inline src/java.base/share/classes/java/lang/invoke/VarHandle.java:2]: Unless this is backport of OpenJDK changes the Oracle copyright shouldn't change. Pls add an IBM copyright statement.

**Comment by pshipton** [comment]
The change order should be head, 22, 21, 17.

**Comment by keithc-ca** [comment]
> The change order should be head, 22, 21, 17.

The head stream should be addressed first (jdknext aka jdk23).

**Comment by jdmpapin** [comment]
I've opened the jdknext PR here: https://github.com/ibmruntimes/openj9-openjdk-jdk/pull/747

**Comment by jdmpapin** [comment]
Yes, I'll have those open momentarily. The changes in those versions are the same as for jdknext

**Comment by jdmpapin** [comment]
I've optimistically opened both:
- https://github.com/ibmruntimes/openj9-openjdk-jdk22/pull/28
- https://github.com/ibmruntimes/openj9-openjdk-jdk21/pull/132

File: src/java.base/share/classes/java/lang/invoke/VarHandle.java [hunk 2]
```diff
@@ -2132,6 +2139,8 @@ public Optional<VarHandleDesc> describeConstable() {
     @Stable
     TypesAndInvokers typesAndInvokers;
 
+    private static final long TYPES_AND_INVOKERS_OFFSET;
+
```
Comment (keithc-ca) [inline src/java.base/share/classes/java/lang/invoke/VarHandle.java:2143]: Please move this just before the static block that initializes it (like newer versions).

**Review by keithc-ca — CHANGES_REQUESTED** [review]

**Review by keithc-ca — APPROVED** [review]

**Comment by keithc-ca** [comment]
Validated locally (similar change included in newer versions).

**Comment by keithc-ca** [comment]
I think a similar change is needed for https://github.com/ibmruntimes/openj9-openjdk-jdk11; could you put that together, please?

**Comment by nbhuiyan** [comment]
The problem that these PRs aimed to solve does not apply in JDK11 for two reasons:
* This requires using OpenJDK MethodHandle/VarHandle implementation, which is not the case in OpenJ9 with JDK11 by default.
* Even if OpenJDK MH/VH were enabled in JDK11, the VarHandle-related optimizations necessary for the race condition to be a problem is only enabled for JDK17+.

**Comment by keithc-ca** [comment]
It is my understanding that we intend to adopt OpenJDK MethodHandle/VarHandle implementation in jdk11 at some point. I think it better to address the known problem now rather than hope we'll remember or notice that a change is needed (because a test fails, or, worse, a customer has a problem).

**Pull request merged**